contracts/stark-proxy/v2/impl/SP2Exchange.sol

```diff
@@ -0,0 +1,275 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity 0.7.5;
+pragma abicoder v2;
+
+import { SafeMath } from '../../../dependencies/open-zeppelin/SafeMath.sol';
+import { IERC20 } from '../../../interfaces/IERC20.sol';
+import { IStarkPerpetual } from '../../../interfaces/IStarkPerpetual.sol';
+import { SP1Balances } from '../../v1/impl/SP1Balances.sol';
+
+/**
+ * @title SP2Exchange
+ * @author dYdX
+ *
+ * @dev Handles calls to the StarkPerpetual contract, for interacting with the dYdX L2 exchange.
+ *
+ *  Standard exchange operation is handled by EXCHANGE_OPERATOR_ROLE. The “forced” actions can only
+ *  be called by the OWNER_ROLE or GUARDIAN_ROLE. Some other functions are also callable by
+ *  the GUARDIAN_ROLE.
+ *
+ *  See SP1Roles, SP2Guardian, SP2Owner, and SP2Withdrawals.
+ */
+abstract contract SP2Exchange is
+  SP1Balances
+{
+  using SafeMath for uint256;
+
+  // ============ Constants ============
+
+  IStarkPerpetual public immutable STARK_PERPETUAL;
+
+  // ============ Events ============
+
+  event DepositedToExchange(
+    uint256 starkKey,
+    uint256 starkAssetType,
+    uint256 starkVaultId,
+    uint256 tokenAmount
+  );
+
+  event WithdrewFromExchange(
+    uint256 starkKey,
+    uint256 starkAssetType,
+    uint256 tokenAmount,
+    bool isGuardianAction
+  );
+
+  /// @dev Limited fields included. Details can be retrieved from Starkware logs if needed.
+  event RequestedForcedWithdrawal(
+    uint256 starkKey,
+    uint256 vaultId,
+    bool isGuardianAction
+  );
+
+  /// @dev Limited fields included. Details can be retrieved from Starkware logs if needed.
+  event RequestedForcedTrade(
+    uint256 starkKey,
+    uint256 vaultId,
+    bool isGuardianAction
+  );
+
+  /// @dev Limited fields included. Details can be retrieved from Starkware logs if needed.
+  event DepositCanceled(
+    uint256 starkKey,
+    uint256 vaultId,
+    bool isGuardianAction
+  );
+
+  /// @dev Limited fields included. Details can be retrieved from Starkware logs if needed.
```

Comment:
I think let's get rid of these warnings on `DepositCanceled` and `DepositReclaimed` and make the params consistent with `DepositedToExchange` and `WithdrewFromExchange`:
* add `uint256 starkAssetType` to both
* add `uint256 tokenAmount` to `DepositReclaimed` since it involves a token transfer